node/node_cross_shard.go

```diff
@@ -120,124 +119,75 @@ func (node *Node) VerifyBlockCrossLinks(block *types.Block) error {
 
-		headersToQuque := []*block.Header{}
+		candidates := []types.CrossLink{}
+		utils.Logger().Debug().
+			Msgf("[ProcessingCrossLink] Crosslink going to propose: %d", len(crosslinks))
 
-		for _, header := range crossLinkHeadersToProcess {
-			if len(headersToQuque) > crossLinkBatchSize {
-				break
+		for i, cl := range crosslinks {
+			if cl.Number() == nil || cl.Number().Cmp(firstCrossLinkBlock) < 0 {
+				utils.Logger().Debug().
+					Msgf("[ProcessingCrossLink] Crosslink %d skipped: %v", i, cl)
+				continue
 			}
-			exist, err := node.Blockchain().ReadCrossLink(header.ShardID(), header.Number().Uint64(), false)
+			exist, err := node.Blockchain().ReadCrossLink(cl.ShardID(), cl.Number().Uint64())
 			if err == nil && exist != nil {
 				utils.Logger().Debug().
-					Msgf("[ProcessingHeader] Cross Link already exists, pass. Block num: %d, shardID %d", header.Number(), header.ShardID())
+					Msgf("[ProcessingCrossLink] Cross Link already exists, pass. Block num: %d, shardID %d", cl.Number(), cl.ShardID())
 				continue
 			}
 
-			if header.Number().Cmp(firstCrossLinkBlock) > 0 { // Directly trust the first cross-link
-				// Sanity check on the previous link with the new link
-				previousLink, err := node.Blockchain().ReadCrossLink(header.ShardID(), header.Number().Uint64()-1, false)
-				if err != nil {
-					previousLink, err = node.Blockchain().ReadCrossLink(header.ShardID(), header.Number().Uint64()-1, true)
-					if err != nil {
-						headersToQuque = append(headersToQuque, header)
-						utils.Logger().Error().Err(err).
-							Msgf("[ProcessingHeader] ReadCrossLink cannot read previousLink with number %d, shardID %d", header.Number().Uint64()-1, header.ShardID())
-						continue
-					}
-				}
-
-				err = node.VerifyCrosslinkHeader(previousLink.Header(), header)
-				if err != nil {
-					utils.Logger().Error().
-						Err(err).
-						Msgf("[ProcessingHeader] Failed to verify new cross link header for shardID %d, blockNum %d", header.ShardID(), header.Number())
-					continue
-				}
+			err = node.VerifyCrossLink(cl)
+			if err != nil {
+				utils.Logger().Error().
+					Err(err).
+					Msgf("[ProcessingCrossLink] Failed to verify new cross link for shardID %d, blockNum %d", cl.ShardID(), cl.Number())
+				continue
 			}
-
-			crossLink := types.NewCrossLink(header)
+			candidates = append(candidates, cl)
 			utils.Logger().Debug().
-				Msgf("[ProcessingHeader] committing for shardID %d, blockNum %d", header.ShardID(), header.Number().Uint64())
-			node.Blockchain().WriteCrossLinks(types.CrossLinks{crossLink}, true)
+				Msgf("[ProcessingCrossLink] committing for shardID %d, blockNum %d", cl.ShardID(), cl.Number().Uint64())
 		}
-
-		// Queue up the cross links that's in the future
-		node.pendingClMutex.Lock()
-		node.pendingCrossLinks = append(node.pendingCrossLinks, headersToQuque...)
-		node.pendingClMutex.Unlock()
+		node.pendingCLMutex.Lock()
+		node.pendingCrossLinks = append(node.pendingCrossLinks, candidates...)
```

Comment:
We still need to put cross link into the DB before it's proposed in blockchain. If a node is down, the in-memory data is all gone.

Reply:
use local db instead of in-memory